ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -271,4 +489,376 @@ private void updateMapHelper(Map<String, Long> map, String key, Long value) {
+    long startTimeInMs;
+
+    @Override
+    public void run() {
+      try {
+        if (cancelled) {
+          return;
+        }
+        recentEntryQueueEnabled = false;
+        recentEntryQueue.clear();
+        queueEntryCount.set(0);
+        metrics.statsRecentEntryQueueSize.update(0);
+        startTimeInMs = time.milliseconds();
+        newScanResults = new ScanResults(startTimeInMs, logSegmentForecastOffsetMs, bucketCount, bucketSpanInMs);
+        scanLock.lock();
+        isScanning = true;
+        scanLock.unlock();
+        Offset firstCheckpoint = index.getCurrentEndOffset();
+        ConcurrentNavigableMap<Offset, IndexSegment> indexSegments = index.getIndexSegments();
+        Map<StoreKey, Long> deletedKeys = new HashMap<>();
+        // process the active index segment based on the firstCheckpoint in case index segment rolled over after the
+        // checkpoint is taken
+        if (indexSegments.size() > 0) {
+          Map.Entry<Offset, IndexSegment> activeIndexSegmentEntry = indexSegments.floorEntry(firstCheckpoint);
+          long indexSegmentStartProcessTime = time.milliseconds();
+          List<IndexEntry> activeIndexEntries =
+              getFilteredIndexEntries(activeIndexSegmentEntry.getValue(), firstCheckpoint);
+          processIndexSegmentEntriesBackward(activeIndexSegmentEntry.getValue(), activeIndexEntries, deletedKeys);
+          metrics.statsBucketingScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTime,
+              TimeUnit.MILLISECONDS);
+          if (indexSegments.size() > 1) {
+            ConcurrentNavigableMap<Offset, IndexSegment> sealedIndexSegments =
+                indexSegments.subMap(index.getStartOffset(), activeIndexSegmentEntry.getKey());
+            for (IndexSegment indexSegment : sealedIndexSegments.descendingMap().values()) {
+              if (cancelled) {
+                return;
+              }
+              indexSegmentStartProcessTime = time.milliseconds();
+              List<IndexEntry> indexEntries = getIndexEntries(indexSegment);
+              processIndexSegmentEntriesBackward(indexSegment, indexEntries, deletedKeys);
+              metrics.statsBucketingScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTime,
+                  TimeUnit.MILLISECONDS);
+            }
+          }
+        } else {
+          newScanResults.updateLogSegmentBaseBucket(index.getStartOffset().getName(), 0L);
+        }
+        recentEntryQueueEnabled = true;
+        Offset secondCheckpoint = index.getCurrentEndOffset();
+        if (secondCheckpoint.compareTo(firstCheckpoint) > 0) {
+          forwardScan(firstCheckpoint, secondCheckpoint);
+        }
+        newScanResults.scannedEndOffset = secondCheckpoint;
+        scanResults.set(newScanResults);
+      } catch (Exception e) {
+        logger.error("Exception while scanning index for stats bucketing", e);
+        metrics.blobStoreStatsErrorCount.inc();
```

Comment:
The blobStoreStatsErrorCount is called multiple times but there are only a few error types. I will split them into errors cause by IndexScanner and error caused by RecentEntryQueue.